Implement initial version of Hypervisor crate (abstraction between KVM and Microsoft Hyper-V)

russell-islam opened the pull request:
This is the initial implementation of the proposal that we made earlier. The purpose of this crate is to make the hypervisor support abstracted to include Microsoft Hyper-V. We will have a hypervisor trait that will be separately implemented by KVM and Hyper-V. Based on the hypervisor we will create the VMfs and VcpuFD which are also some genetic traits. These two traits will be implemented by KVM and HyperV. For KVM there will be two implementations KvmVMfd and KvmVcpuFd and same for HyperV.

Sample code: hypervisor = KvmHyperVisor::new()
vm = hypervisor.create_vm()
vcpu = vm.create_vcpu(id)

vm.get/set()
vcpu.get/set()

For the payload we have a binding inside the hypervisor crate that will abstract between HyperV and KVM.

The plan is to move most of the KVM reference to Hypervisor crate. The hypervisor will hand the payload based on KVM or HyperV. This is the initial proposal. For the payload I have followed the previous commit by crowd-strike.

russell-islam commented:
Please don't merge!!

russell-islam commented:
Looks like build on aarch64 fails. Is there anyway to build aarch64 binary on x86-64 machine?

michael2012z commented:
@russell-islam , you can **cross-build** for aarch64 on x86_64, that's what we do with the 2 "Cross Build" checks now.

To try cross-build locally, you need to install target `aarch64-unknown-linux-gnu` or `aarch64-unknown-linux-musl`, and a tool-chain (to link).

michael2012z commented:
About the build error on aarch64.  It was caused by an issue in `cloud-hypervisor/kvm-bindings` on aarch64, see this PR for the complete story: https://github.com/cloud-hypervisor/kvm-bindings/pull/1.

As a workaround, you can modify the `Cargo.toml` (which uses `cloud-hypervisor/kvm-bindings` with `with-serde` feature) when it comes to aarch64 build, here is an example: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/.github/workflows/cross-build.yaml#L26

yisun-git commented:
Hi,

We proposed the Hypervisor crate in cloud-hypervisor and rust-vmm before.
https://github.com/rust-vmm/community/issues/50
https://github.intel.com/cloud-hypervisor/community/issues/2

The draft codes had been provided at:
https://github.com/yisun-git/Hypervisor_on_FC/tree/dbg_yisun1_hypervisor

But the community has different opinions on this. So, we stopped the work. 

I want to learn what is the purpose of this PR? Have you addressed the comments in above issues?

Thanks,
Sun Yi

russell-islam commented:
> @russell-islam , you can **cross-build** for aarch64 on x86_64, that's what we do with the 2 "Cross Build" checks now.
> 
> To try cross-build locally, you need to install target `aarch64-unknown-linux-gnu` or `aarch64-unknown-linux-musl`, and a tool-chain (to link).

Is this documented somewhere with detailed steps?

russell-islam commented:
> Hi,
> 
> We proposed the Hypervisor crate in cloud-hypervisor and rust-vmm before.
> [rust-vmm/community#50](https://github.com/rust-vmm/community/issues/50)
> https://github.intel.com/cloud-hypervisor/community/issues/2
> 
> The draft codes had been provided at:
> https://github.com/yisun-git/Hypervisor_on_FC/tree/dbg_yisun1_hypervisor
> 
> But the community has different opinions on this. So, we stopped the work.
> 
> I want to learn what is the purpose of this PR? Have you addressed the comments in above issues?
> 
> Thanks,
> Sun Yi

@yisun-git  Thanks for the information. The purpose of this PR is to get initial comments on the Hypervisor crate that will eventually support Microsoft Hyper-V.

russell-islam commented:
> Hi,
> 
> We proposed the Hypervisor crate in cloud-hypervisor and rust-vmm before.
> [rust-vmm/community#50](https://github.com/rust-vmm/community/issues/50)
> https://github.intel.com/cloud-hypervisor/community/issues/2
> 
> The draft codes had been provided at:
> https://github.com/yisun-git/Hypervisor_on_FC/tree/dbg_yisun1_hypervisor
> 
> But the community has different opinions on this. So, we stopped the work.
> 
> I want to learn what is the purpose of this PR? Have you addressed the comments in above issues?
> 
> Thanks,
> Sun Yi

@yisun-git  Thanks for the information. The purpose of this PR is to get initial comments on the Hypervisor crate that will eventually support Microsoft Hyper-V.

russell-islam commented:
Aarch64 build is fine now but I am getting the following now. Looks like  this a known issue even with latest master:
`error[E0277]: the trait bound `kvm_bindings::x86::bindings_v4_20_0::kvm_sregs: serde::Serialize` is not satisfied`

michael2012z commented on vmm/Cargo.toml:
```diff
@@ -20,7 +20,7 @@ anyhow = "1.0"
 devices = { path = "../devices" }
 epoll = ">=4.0.1"
-kvm-bindings = { git = "https://github.com/cloud-hypervisor/kvm-bindings", branch = "ch", features = ["with-serde", "fam-wrappers"] }
+kvm-bindings = { git = "https://github.com/cloud-hypervisor/kvm-bindings", branch = "ch", features = ["fam-wrappers"] }
```
Now the problem is: `"with-serde"` is needed on x86_64, but not on aarch64. 
So don't modify this, so x86_64 should build.
For aarch64, replace https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/.github/workflows/cross-build.yaml#L26 with `run: sed -i 's/"with-serde",\ //g' vmm/Cargo.toml && sed -i 's/"with-serde",\ //g' hypervisor/Cargo.toml` in your PR. Then both x86_64 and  aarch64 should build.

michael2012z commented on hypervisor/Cargo.toml:
```diff
@@ -13,7 +13,7 @@ hyperv = []
 devices = { path = "../devices" }
 kvm-ioctls = { git = "https://github.com/cloud-hypervisor/kvm-ioctls", branch = "ch" }
-kvm-bindings = { git = "https://github.com/cloud-hypervisor/kvm-bindings", branch = "ch", features = ["with-serde", "fam-wrappers"] }
+kvm-bindings = { git = "https://github.com/cloud-hypervisor/kvm-bindings", branch = "ch", features = ["fam-wrappers"] }
```
Same as https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1225/commits/e3129de507f83e479207da93692342f60bf07151#r430818677
Don't remove `"with-serde"`.

russell-islam commented:
@michael2012z Now the build fails with following
`Compiling kvm-bindings v0.2.0 (https://github.com/cloud-hypervisor/kvm-bindings?branch=ch#3a678008)
error: cannot find derive macro `Deserialize` in this scope
  --> /cargo/git/checkouts/kvm-bindings-c6b55506be61a061/3a67800/src/arm64/mod.rs:29:43
   |
29 | #[cfg_attr(feature = "with-serde", derive(Deserialize, Serialize))]
   |                                           ^^^^^^^^^^^

error: cannot find derive macro `Serialize` in this scope
  --> /cargo/git/checkouts/kvm-bindings-c6b55506be61a061/3a67800/src/arm64/mod.rs:29:56
   |
29 | #[cfg_attr(feature = "with-serde", derive(Deserialize, Serialize))]
   |                                                        ^^^^^^^^^

   Compiling devices v0.1.0 (/project/devices)
   Compiling net_util v0.1.0 (/project/net_util)
error: aborting due to 2 previous errors

error: could not compile `kvm-bindings`.

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
`

michael2012z commented:
@russell-islam , see https://github.com/michael2012z/cloud-hypervisor/pull/28, attention to the last commit.

russell-islam commented:
> @russell-islam , see [michael2012z#28](https://github.com/michael2012z/cloud-hypervisor/pull/28), attention to the last commit.

Thank you. I misunderstood the fix. Now I got it.

sameo commented on hypervisor/src/kvm.rs:
```diff
@@ -0,0 +1,273 @@
+        self.fd.set_gsi_routing(irq_routing)
+    }
+    fn set_user_memory_region(&self, user_memory_region: MemoryRegion) -> ResultOps<()> {
+        unsafe { self.fd.set_user_memory_region(user_memory_region) }
```
For unsafe blocks, we should add a comment describing why we think it is still safe.

sameo commented on vfio/src/vfio_pci.rs:
```diff
@@ -595,7 +595,7 @@ impl VfioPciDevice {
                     flags: 0,
                 };
                 // Safe because the guest regions are guaranteed not to overlap.
```
The comment could be removed.

sameo commented on vmm/src/cpu.rs:
```diff
@@ -527,7 +450,6 @@ impl Vcpu {
     #[cfg(target_arch = "x86_64")]
     fn set_kvm_state(&mut self, state: &VcpuKvmState) -> Result<()> {
         self.fd.set_regs(&state.regs).map_err(Error::VcpuSetRegs)?;
-
```
Unneeded change.

sameo commented on hypervisor/src/lib.rs:
```diff
@@ -0,0 +1,113 @@
+    Err(Error::HyperVisorTypeMismatch)
+}
+
+pub trait GenVcpuFd: Send + Sync {
```
It would be cleaner if we could separate the traits into separate files, on for the hypervisor, and another one for the vcpu trait.

sameo replied:
Just calling that one `Vcpu` would be nicer.

sameo commented on hypervisor/src/cpuidpatch.rs:
```diff
@@ -0,0 +1,137 @@
+    });
+
+    // Supported CPUID
+    let mut cpuid = kvm
+        .get_supported_cpuid(kvm_bindings::KVM_MAX_CPUID_ENTRIES)
+        .map_err(Error::PatchCpuId)?;
+
```
This part of the file is basically the only one that's really hypervisor specific, so I think it would  methomake more sense to abstract a `get_cpuid` method through the `Vcpu` trait, and keep the whole cpuid patching implementation under vmm/src/cpu.rs.
Any other reason why you moved it here?

russell-islam commented:
Thanks a lot @sameo . I will address your comments.

sboeuf commented on Cargo.toml:
```diff
@@ -11,6 +11,7 @@ arc-swap = ">=0.4.4"
 clap = { version = "2.33.1", features=["wrap_help"] }
 epoll = ">=4.0.1"
 futures = { version = "0.3.5", features = ["thread-pool"] }
+hypervisor = { path = "hypervisor" }
```
Could you please add an extra commit introducing the new `hypervisor` crate without anything inside?
This commit should be the first of the entire patchset, and it will be the opportunity to describe the goal of this new crate in the commit message.

russell-islam replied:
@sameo I guess I can add one more extra commit just to add the skeleton of the hypervisor crate. vcpu would be the next commit.

sboeuf commented on hypervisor/src/kvm.rs:
```diff
@@ -0,0 +1,1269 @@
+    fn set_tss_address(&self, offset: usize) -> vm::Result<()> {
+        self.fd
+            .set_tss_address(offset)
+            .map_err(|e| vm::HypervisorVmError::SetTssAddress(e.into()))
```
I think you can simplify every
```rust
map_err(|e| vm::HypervisorVmError::SpecificError(e.into()))
```
with
```rust
map_err(vm::HypervisorVmError::SpecificError)
```

russell-islam replied:
If I change this I get the following error. At first I was trying to keep it kvm_ioctls::errorno::Error but in order to make it hypervisor agnostic Samuel suggested to use the current implementation.
`SetTssAddress(#[source] anyhow::Error),
TSS address error

type mismatch in function arguments
expected signature of `fn(vmm_sys_util::errno::Error) -> _`rustc(E0631)
vm.rs(31, 5): found signature of `fn(anyhow::Error) -> _`

sboeuf replied:
Oh I see. You need to convert the `kvm_ioctls::errorno::Error` error that you get from `kvm-ioctls` into `anyhow::Error`, so that from the trait we specify `anyhow::Error` only.

sboeuf commented on hypervisor/src/kvm.rs:
```diff
@@ -0,0 +1,1269 @@
+    ///             exit_reason => panic!("unexpected exit reason: {:?}", exit_reason),
+    ///         }
+    ///     }
+    /// }
```
I'm wondering if all these examples/documentation are needed since they've been ported from the `kvm-ioctls` crate. That feels like a lot of duplicates.

russell-islam replied:
Right. But the example has some changes reflecting the current hypervisor implementations.

sboeuf replied:
Right, but could we keep it shorter since the core content is already part of `kvm-ioctls`?

russell-islam replied:
Will do . thanks

russell-islam commented:
I will make the change once I get some reviews to avoid back and forth update.

sameo commented on hypervisor/src/common.rs:
```diff
@@ -16,3 +16,8 @@ pub use {
     kvm_bindings::kvm_vcpu_events as VcpuEvents, kvm_ioctls::DeviceFd, kvm_ioctls::IoEventAddress,
     kvm_ioctls::VcpuExit,
 };
+
```
It would be good to amend the commit message as to briefly describe why this is needed.

sameo commented on arch/src/x86_64/mod.rs:
```diff
@@ -15,8 +15,7 @@ mod mptable;
 use crate::RegionType;
-use kvm_bindings::CpuId;
-use kvm_ioctls::*;
+
```
Unneeded change.

sameo commented on arch/src/x86_64/interrupts.rs:
```diff
@@ -12,7 +12,7 @@ use std::sync::Arc;
 use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
 
-use kvm_bindings::kvm_lapic_state;
+use hypervisor::kvm_lapic_state;
```
This should be `use hypervisor::x86_64::LapicState` and then replace all `kvm_lapic_state` instances with `LapicState`

sameo commented on arch/src/x86_64/mod.rs:
```diff
@@ -15,8 +15,7 @@ mod mptable;
 use crate::InitramfsConfig;
 use crate::RegionType;
-use kvm_bindings::CpuId;
-use kvm_ioctls::*;
```
`use hypervisor::CpuId` would remove the need for some of the below changes.

sameo commented on arch/src/x86_64/gdt.rs:
```diff
@@ -8,7 +8,7 @@
 
 // For GDT details see arch/x86/include/asm/segment.h
-use kvm_bindings::kvm_segment;
+use hypervisor::kvm_segment;
```
Same here, you want to use `hypervisor::x86_64::SegmentRegister`

sameo commented on hypervisor/src/lib.rs:
```diff
@@ -39,7 +39,17 @@ pub mod kvm;
+#[cfg(target_arch = "x86_64")]
+pub use kvm_bindings::{kvm_lapic_state, kvm_segment, kvm_userspace_memory_region};
+pub use kvm_ioctls;
+pub use kvm_ioctls::VcpuExit;
```
This should be exported through the kvm module only.

russell-islam commented on hypervisor/src/vm.rs:
```diff
@@ -0,0 +1,110 @@
+    /// Registers an event to be signaled whenever a certain address is written to.
+    fn register_ioevent(
+        &self,
+        fd: &EventFd,
```
Yes. I totally agree. We have to adjust the trait once we have hyperv-ioctls ready.

russell-islam commented on hypervisor/src/x86_64/mod.rs:
```diff
@@ -0,0 +1,15 @@
+pub use {
+    kvm_bindings::kvm_lapic_state as LapicState,
+    kvm_bindings::kvm_xcrs as ExtendedControlRegisters, kvm_bindings::kvm_xsave as Xsave,
+    kvm_bindings::CpuId, kvm_bindings::Msrs as MsrEntries, kvm_ioctls::Cap, kvm_ioctls::Kvm,
```
I need to remove some vVsCode code fmt. This is auto saved. :(

russell-islam replied:
This is not accepted by cargo fmt check. I had to revert it. :(

russell-islam commented on hypervisor/src/cpu.rs:
```diff
@@ -0,0 +1,229 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
```
I wile change this to follwoing:
`// Copyright © 2019 Intel Corporation
//
// SPDX-License-Identifier: Apache-2.0
//
// Copyright © 2020, Microsoft  Corporation
//
// Copyright 2018-2019 CrowdStrike, Inc.
//`

russell-islam replied:
Guru is in a talk within Msft. I will update you once I hear something.

russell-islam replied:
@sameo We are fine with dual licensing apache or MIT

sameo commented on hypervisor/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+
+[dependencies.linux-loader]
+git = "https://github.com/rust-vmm/linux-loader"
+features = ["elf", "bzimage"]
```
I cleaned up the cargo.toml dependencies, so that there is no cloud-hypervisor crate dependencies from this crate. See https://github.com/russell-islam/cloud-hypervisor/pull/1/commits/1d5c64bc02c10cc8d01925a8ee45ffd9ed9f9585

russell-islam commented on vmm/src/vm.rs:
```diff
@@ -379,8 +335,9 @@ impl Vm {
+        hypervisor: Arc<dyn hypervisor::Hypervisor>,
     ) -> Result<Self> {
-        let (kvm, fd) = Vm::kvm_new()?;
+        let fd = hypervisor.create_vm().unwrap();
```
This is not right. I will address this one.

russell-islam commented on hypervisor/src/kvm.rs:
```diff
@@ -0,0 +1,464 @@
+        }
+        let vm_fd = Arc::new(fd);
+
+        // Set TSS
+        #[cfg(target_arch = "x86_64")]
+        vm_fd
+            .set_tss_address(KVM_TSS_ADDRESS.raw_value() as usize)
+            .map_err(|e| hv::HypervisorError::VmSetup(e.into()))?;
+
+        #[cfg(target_arch = "x86_64")]
+        {
+            // Create split irqchip
+            // Only the local APIC is emulated in kernel, both PICs and IOAPIC
+            // are not.
+            let mut cap: kvm_enable_cap = Default::default();
+            cap.cap = KVM_CAP_SPLIT_IRQCHIP;
+            cap.args[0] = ioapic::NUM_IOAPIC_PINS as u64;
+            vm_fd
+                .enable_cap(&cap)
+                .map_err(|e| hv::HypervisorError::VmSetup(e.into()))?;
+        }
```
We should remove check_required_kvm_extensions from here(craete_vm) as well. right?

sboeuf replied:
Yes.

russell-islam replied:
@sboeuf Could you please revisit this?

likebreath commented:
Hi @russell-islam, following our conversation last Friday, I got some updates on how you may debug the CI failures locally. I asked inputs from @sboeuf today. He pointed that the CI is failing on (almost) all tests when the "mmio" feature is enabled both for glibc and musl, e.g. the bare-minimum test `test_bzimage_boot`. So basically, there were something fundamentally broken related to "mmio" feature.

I managed to reproduce the failure locally without the integration tests environment. With the following command-line to launch cloud-hypervisor, I got [master branch](a17aaec663e37aed1e79d0b7b2a1c46cee8de390) booting correctly while your PR branch hanging. Please let me know whether you can reproduce it on your side. If you do, it should give you a good starting point to debug the issue. Any other questions, please let us know. Thank you again for the contribution.

```
$ cargo build  --no-default-features --features "mmio"  
$ sudo setcap cap_net_admin+ep /path/to/cloud-hypervisor/target/debug/cloud-hypervisor 
$ /home/chenb/project/cloud-hypervisor/cloud-hypervisor/target/debug/cloud-hypervisor \                                                                                                                            
    --kernel /path/to/workload/bzImage \                                                                                                                                                                     
    --disk path=/path/to/clear-31890-kvm.img \                                                                                                                
    --cmdline "root=/dev/vda3 \                                                                                                                                                                                  
                     console=tty0 console=ttyS0,115200n8 console=hvc0 quiet \                                                                                                                                    
                     init=/usr/lib/systemd/systemd-bootchart initcall_debug tsc=reliable \                                                                                                                       
                     no_timer_check noreplace-smp cryptomgr.notests \                                                                                                                                            
                     rootfstype=ext4,btrfs,xfs kvm-intel.nested=1 rw" \                                                                                                                                          
    --cpus boot=4 --memory size=1024M  --net "tap=,mac=,ip=,mask=" --rng --api-socket /tmp/api -vv --log-file /tmp/dbg.log   
```

russell-islam commented:
Looks like mmio feature hangs indefinitely in the vmm_thread.join function call.

likebreath commented:
@russell-islam I played your PR with GDB and took a reference from the working master branch. It looks like the VMM (cloud-hypervisor) is not hanging: vmm/http-servier/vcpu/virtio threads are working fine. What is hanging seems to be the guest which is stuck at the early stage of booting. I also see the following error from the clh debug log, which are MMIO access errors. It aligns with our hypothesis that the problem is tied to MMIO. I will take a closer look tomorrow. Anything you found interesting, please just share/ask.

```
  1cloud-hypervisor: 6.049099ms: INFO:vmm/src/memory_manager.rs:732 -- Created userspace mapping: 0 -> 7fffa8000000 40000000                                                                                       
  2cloud-hypervisor: 683.317154ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
  3cloud-hypervisor: 683.379447ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
  4cloud-hypervisor: 683.409303ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
  5cloud-hypervisor: 683.438271ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
  6cloud-hypervisor: 683.467305ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
  7cloud-hypervisor: 683.496437ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
  8cloud-hypervisor: 683.525402ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
  9cloud-hypervisor: 683.554339ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
 10cloud-hypervisor: 683.589732ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
 11cloud-hypervisor: 683.618777ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50
...
```

russell-islam commented:
> @russell-islam I played your PR with GDB and took a reference from the working master branch. It looks like the VMM (cloud-hypervisor) is not hanging: vmm/http-servier/vcpu/virtio threads are working fine. What is hanging seems to be the guest which is stuck at the early stage of booting. I also see the following error from the clh debug log, which are MMIO access errors. It aligns with our hypothesis that the problem is tied to MMIO. I will take a closer look tomorrow. Anything you found interesting, please just share/ask.
> 
> ```
>   1cloud-hypervisor: 6.049099ms: INFO:vmm/src/memory_manager.rs:732 -- Created userspace mapping: 0 -> 7fffa8000000 40000000                                                                                       
>   2cloud-hypervisor: 683.317154ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>   3cloud-hypervisor: 683.379447ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>   4cloud-hypervisor: 683.409303ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>   5cloud-hypervisor: 683.438271ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>   6cloud-hypervisor: 683.467305ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>   7cloud-hypervisor: 683.496437ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>   8cloud-hypervisor: 683.525402ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>   9cloud-hypervisor: 683.554339ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>  10cloud-hypervisor: 683.589732ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50                                                                                            
>  11cloud-hypervisor: 683.618777ms: WARN:vm-virtio/src/transport/mmio.rs:382 -- unknown virtio mmio register write: 0x50
> ...
> ```

I also saw this error but i thought it was warning. I tried to find the backtrace on th hang. looks like vcpu threads are hanging in the run ioctls call? I am not sure if that is the reason.

`(gdb) info thread
  Id   Target Id         Frame 
  1    Thread 0x7fc027b3ba00 (LWP 5593) "cloud-hyperviso" 0x00007fc027313d2d in __GI___pthread_timedjoin_ex (threadid=140463257827072, thread_return=0x0, abstime=0x0, block=<optimized out>)
    at pthread_join_common.c:89
  2    Thread 0x7fc026963700 (LWP 5594) "vmm" 0x00007fc026e23bb7 in epoll_wait (epfd=10, events=0x7fc020000bc0, maxevents=100, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  3    Thread 0x7fc026762700 (LWP 5595) "http-server" 0x00007fc026e23bb7 in epoll_wait (epfd=13, events=0x7fc018000ee0, maxevents=10, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  4    Thread 0x7fc026561700 (LWP 5615) "vcpu0" 0x00007fc026e185d7 in ioctl () at ../sysdeps/unix/syscall-template.S:78
  5    Thread 0x7fc026360700 (LWP 5616) "vcpu1" 0x00007fc026e185d7 in ioctl () at ../sysdeps/unix/syscall-template.S:78
  6    Thread 0x7fc02615f700 (LWP 5617) "vcpu2" 0x00007fc026e185d7 in ioctl () at ../sysdeps/unix/syscall-template.S:78
  7    Thread 0x7fc025f5e700 (LWP 5618) "vcpu3" 0x00007fc026e185d7 in ioctl () at ../sysdeps/unix/syscall-template.S:78
* 8    Thread 0x7fc025d5d700 (LWP 5622) "signal_handler" 0x00007fc02731c82e in __libc_recv (fd=64, buf=0x7fc025d5c26c, len=1024, flags=0) at ../sysdeps/unix/sysv/linux/recv.c:28
  9    Thread 0x7fc025b5c700 (LWP 5795) "virtio_console" 0x00007fc026e23bb7 in epoll_wait (epfd=73, events=0x7fbfc0000b80, maxevents=100, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  10   Thread 0x7fc02595b700 (LWP 5796) "virtio_rng" 0x00007fc026e23bb7 in epoll_wait (epfd=80, events=0x7fbfc4000b80, maxevents=100, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  11   Thread 0x7fc02575a700 (LWP 5797) "virtio_blk" 0x00007fc026e23bb7 in epoll_wait (epfd=81, events=0x7fbfb8000b80, maxevents=100, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
(gdb) thread 7
[Switching to thread 7 (Thread 0x7fc025f5e700 (LWP 5618))]
#0  0x00007fc026e185d7 in ioctl () at ../sysdeps/unix/syscall-template.S:78
78	../sysdeps/unix/syscall-template.S: No such file or directory.
(gdb) bt
#0  0x00007fc026e185d7 in ioctl () at ../sysdeps/unix/syscall-template.S:78
#1  0x000055fb2d2a9860 in vmm_sys_util::linux::ioctl::ioctl (fd=0x7fc0200063e0, req=44672) at /home/muminul/.cargo/registry/src/github.com-1ecc6299db9ec823/vmm-sys-util-0.6.0/src/linux/ioctl.rs:246
#2  0x000055fb2d2a061d in kvm_ioctls::ioctls::vcpu::VcpuFd::run (self=0x7fc0200063e0) at /home/muminul/.cargo/git/checkouts/kvm-ioctls-545a2ef5ec405886/5b7f1cc/src/ioctls/vcpu.rs:1100
#3  0x000055fb2d281347 in <hypervisor::kvm::KvmVcpu as hypervisor::cpu::Vcpu>::run (self=0x7fc0200063e0) at hypervisor/src/kvm/mod.rs:489
#4  0x000055fb2ccdbef4 in vmm::cpu::Vcpu::run (self=0x7fc020006450) at vmm/src/cpu.rs:297
#5  0x000055fb2cd59fcc in vmm::cpu::CpuManager::start_vcpu::{{closure}} () at vmm/src/cpu.rs:742
#6  0x000055fb2cdd80a0 in std::sys_common::backtrace::__rust_begin_short_backtrace (f=...)
    at /home/muminul/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/sys_common/backtrace.rs:130
#7  0x000055fb2cbd5901 in std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} ()
    at /home/muminul/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/thread/mod.rs:475
#8  0x000055fb2ccebad3 in <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once (self=..., _args=())
    at /home/muminul/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:318
#9  0x000055fb2cde9034 in std::panicking::try::do_call (data=0x7fc025f5d938 "P8") at /home/muminul/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:297
#10 0x000055fb2cde9a2d in __rust_try ()
#11 0x000055fb2cde83d4 in std::panicking::try (f=...) at /home/muminul/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:274
#12 0x000055fb2ccebd43 in std::panic::catch_unwind (f=...) at /home/muminul/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:394
#13 0x000055fb2cbd3f4a in std::thread::Builder::spawn_unchecked::{{closure}} () at /home/muminul/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/thread/mod.rs:474
#14 0x000055fb2cc6819f in core::ops::function::FnOnce::call_once{{vtable-shim}} ()
    at /home/muminul/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/function.rs:232
#15 0x000055fb2d2fb0aa in <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once () at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/liballoc/boxed.rs:1034
#16 <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once () at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/liballoc/boxed.rs:1034
#17 std::sys::unix::thread::Thread::new::thread_start () at src/libstd/sys/unix/thread.rs:87
#18 0x00007fc0273126db in start_thread (arg=0x7fc025f5e700) at pthread_create.c:463
#19 0x00007fc026e2388f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 
`

russell-islam commented:
@sboeuf @rbradford  any input here?

russell-islam commented:
> unknown virtio mmio register write

Thanks @likebreath. Looks like the offset actually notifying the queue index

QueueNotify 0x050 W | Queue notifier Writing a queue index to this register notifies the device   that there are new buffers to process in the queue.
-- | --

sameo commented:
@russell-islam You were registering the wrong io events and not getting guest notifications. That:
https://github.com/sameo/cloud-hypervisor-1/commit/c07ff474df496f278bca0396726280f4dbed5a36

should fix it.

likebreath commented:
@russell-islam I tried the patch from @sameo, and now clh with MMIO feature boots correctly. :)

I think Sam's branch is here https://github.com/sameo/cloud-hypervisor-1/commits/topic/hypervisor and the updated patch I tried is here: https://github.com/sameo/cloud-hypervisor-1/commit/cad17ffe110c934a613051a8908765bc452a06d4.

sameo commented:
> @russell-islam I tried the patch from @sameo, and now clh with MMIO feature boots correctly. :)
> 
> I think Sam's branch is here https://github.com/sameo/cloud-hypervisor-1/commits/topic/hypervisor and the updated patch I tried is here: [sameo@cad17ff](https://github.com/sameo/cloud-hypervisor-1/commit/cad17ffe110c934a613051a8908765bc452a06d4).

@likebreath is correct, https://github.com/sameo/cloud-hypervisor-1/commit/cad17ffe110c934a613051a8908765bc452a06d4 is an implementation that's correct, although there may be more idiomatic way to do it.

likebreath commented:
Great. The CI now is passing. Now only build failing with aarch64 w/ musl.

russell-islam commented:
> Great. The CI now is passing. Now only build failing with aarch64 w/ musl.

@likebreath thanks for keeping an aye. The integration tests passed. yeah!!

sameo commented:
@sboeuf Are you happy with the changes? Please have a look. As a first step towards a complete hypervisor absraction, this PR is ready to be reviewed imho.

sboeuf commented on hypervisor/src/kvm/x86_64/mod.rs:
```diff
@@ -16,3 +20,21 @@ pub use {
+///
+/// Check KVM extension for Linux
+///
+pub fn check_required_kvm_extensions(kvm: &Kvm) -> KvmResult<()> {
+    if !kvm.check_extension(Cap::SignalMsi) {
+        return Err(KvmError::CapabilityMissing(Cap::SignalMsi));
+    }
+    if !kvm.check_extension(Cap::TscDeadlineTimer) {
+        return Err(KvmError::CapabilityMissing(Cap::TscDeadlineTimer));
+    }
+    if !kvm.check_extension(Cap::SplitIrqchip) {
+        return Err(KvmError::CapabilityMissing(Cap::SplitIrqchip));
+    }
+    Ok(())
```
This is very Cloud-Hypervisor specific. I mean the requirement for having these KVM caps supported is not something that all VMM might need, so it'd be more appropriate to let the VMM check them.

russell-islam replied:
In this case we need to have a way to return this kvm object reference back to vmm.

sboeuf commented on hypervisor/src/kvm/x86_64/mod.rs:
```diff
@@ -10,19 +10,66 @@
+}
+
+pub fn boot_msr_entries() -> MsrEntries {
+    MsrEntries::from_entries(&[
+        msr!(msr_index::MSR_IA32_SYSENTER_CS),
+        msr!(msr_index::MSR_IA32_SYSENTER_ESP),
+        msr!(msr_index::MSR_IA32_SYSENTER_EIP),
+        msr!(msr_index::MSR_STAR),
+        msr!(msr_index::MSR_CSTAR),
+        msr!(msr_index::MSR_LSTAR),
+        msr!(msr_index::MSR_KERNEL_GS_BASE),
+        msr!(msr_index::MSR_SYSCALL_MASK),
+        msr!(msr_index::MSR_IA32_TSC),
+        msr_data!(
+            msr_index::MSR_IA32_MISC_ENABLE,
+            msr_index::MSR_IA32_MISC_ENABLE_FAST_STRING as u64
+        ),
+        msr_data!(msr_index::MSR_MTRRdefType, MTRR_ENABLE | MTRR_MEM_TYPE_WB),
```
A bit the same idea as the caps here, but I'd like to mention this list of MSRs is specific to Cloud-Hypervisor.

bonzini commented:
Out of curiosity, will the Hypervisor implementation for Hyper-V support Linux as well, or will it be only for Windows (similar to the WHPX support in QEMU)?

sameo commented:
> Out of curiosity, will the Hypervisor implementation for Hyper-V support Linux as well, or will it be only for Windows (similar to the WHPX support in QEMU)?

It will support KVM and Hyper-V on Linux.

vsrivatsa1 commented:
Hello, what's the status of this work being merged into rust-vmm?
